main/src/library/File.flix

```diff
@@ -840,4 +840,71 @@ namespace File {
                 Err(getMessage(ex))
         }
 
+    ///
+    /// Returns a list naming the files and directories in the directory `f`.
+    ///
+    /// Does not recursively traverse the directory.
+    ///
+    pub def list(f: String): Result[List[String], String] & Impure =
+        try {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.list(): Array[String] & Impure;
+
+            let javaFile = newFile(f);
+            let javaList = list(javaFile);
+
+            Ok(Array.toList(javaList))
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage(): String & Impure;
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Returns an iterator naming the files and directories in the directory `f`.
+    ///
+    /// Does not recursively traverse the directory.
+    ///
+    pub def listWithIter(f: String): Result[Iterator[String], String] & Impure =
+        try {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.toPath(): ##java.nio.file.Path & Impure;
+            import static java.nio.file.Files.newDirectoryStream(##java.nio.file.Path): ##java.nio.file.DirectoryStream & Impure;
+            import java.nio.file.DirectoryStream.iterator(): ##java.util.Iterator & Impure;
+
+            import java.util.Iterator.hasNext(): Bool & Impure;
+            import java.util.Iterator.next(): ##java.nio.file.Path & Impure as getNext;
+
+            import java.nio.file.Path.getFileName(): ##java.nio.file.Path & Impure;
+            import java.nio.file.Path.toString(): String & Impure;
+
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let javaStream = newDirectoryStream(javaPath);
+            let javaIter = iterator(javaStream);
+
+            let done = () -> not hasNext(javaIter);
+            let next = () -> {
+                if(hasNext(javaIter)) {
+                    let name = getFileName(getNext(javaIter));
+                    toString(name)
+                } else {
+                    bug!("Empty iterator.")
+                }
+            };
+
+            Ok(Iterator(done, next))
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage(): String & Impure;
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Returns an iterator naming the files and directories in the directory `f`.
+    ///
+    /// Recursively traverses the directory.
+    ///
+    pub def walk(f: String): Result[Iterator[String], String] = ???
```

Comment:
I'm a bit unsure of how to do this recursively. I looked at the [FileVisitor](https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileVisitor.html) interface, but this requires implementing the interface. Searching the web, it does not seem like java have an easy way of just getting `child` of a directory and calling `nextSibling` or something like that. Any suggestions?

Reply:
Could you manually recurse through the folders? By using `list` and then if something is a directory you go into it?

Reply:
Should it name a directory before going into it? Or should it name the directory before?

Reply:
And should it name the whole path or only the name of the file or directory?

Reply:
I just realised, that there is a [walk function in java](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#walk-java.nio.file.Path-java.nio.file.FileVisitOption...-), too.